5.4 regression: C++ compilers fail to compile misc.h

Yet another C++ compatibility issue (see https://github.com/ocaml/ocaml/pull/13777)
```
#=== ERROR while compiling taglib.0.3.10 ======================================#
# context              2.4.0~alpha2 | linux/x86_64 | ocaml-base-compiler.5.4.0~alpha1 | file:///home/opam/opam-repository
# path                 ~/.opam/5.4/.opam-switch/build/taglib.0.3.10
# command              ~/.opam/5.4/bin/dune build -p taglib -j 1 @install
# exit-code            1
# env-file             ~/.opam/log/taglib-14-e79929.env
# output-file          ~/.opam/log/taglib-14-e79929.out
### output ###
# File "src/dune", line 8, characters 9-21:
# 8 |   (names taglib_stubs)
#              ^^^^^^^^^^^^
# (cd _build/default/src && /usr/bin/gcc -fPIC -I/usr/include/taglib -g -I /home/opam/.opam/5.4/lib/ocaml -I /home/opam/.opam/5.4/lib/ocaml/unix -o taglib_stubs.o -c taglib_stubs.cc)
# In file included from /home/opam/.opam/5.4/lib/ocaml/caml/alloc.h:19,
#                  from taglib_stubs.cc:98:
# /home/opam/.opam/5.4/lib/ocaml/caml/misc.h:622:8: error: '_Atomic' does not name a type; did you mean 'Atom'?
#   622 | extern _Atomic uintnat caml_verb_gc;
#       |        ^~~~~~~
#       |        Atom
```

--- a/Changes
+++ b/Changes
@@ -935,6 +935,10 @@ OCaml 5.4.0
   (Stephen Dolan, report by Jan Midtgaard, investigation by Nick Roberts,
    review by Antonin Décimo and Miod Vallat)
 
+- #14101, #14139: define atomic helper types inside `caml/misc.h` to improve
+  header compatibility with C++
+  (Florian Angeletti, report by Kate Deplaix, review by Gabriel Scherer)
+
 - #14135: Fix a rare internal typechecker error when combining recursive
   modules, polymorphic fields or methods, and constrained type parameters.
   (Florian Angeletti, review by Gabriel Scherer)

--- a/runtime/caml/camlatomic.h
+++ b/runtime/caml/camlatomic.h
@@ -25,24 +25,13 @@
  */
 
 #ifdef __cplusplus
-
 extern "C++" {
-#include <atomic>
-typedef std::atomic<uintnat> atomic_uintnat;
-typedef std::atomic<intnat> atomic_intnat;
 using std::memory_order_relaxed;
 using std::memory_order_acquire;
 using std::memory_order_release;
 using std::memory_order_acq_rel;
 using std::memory_order_seq_cst;
 }
-
-#else
-
-#include <stdatomic.h>
-typedef _Atomic uintnat atomic_uintnat;
-typedef _Atomic intnat atomic_intnat;
-
 #endif
 
 #ifdef CAML_INTERNALS

--- a/runtime/caml/misc.h
+++ b/runtime/caml/misc.h
@@ -77,6 +77,17 @@
 /* Basic types and constants */
 
 typedef size_t asize_t;
+#ifdef __cplusplus
+extern "C++" {
+#include <atomic>
+typedef std::atomic<uintnat> atomic_uintnat;
+typedef std::atomic<intnat> atomic_intnat;
+}
+#else
+#include <stdatomic.h>
+typedef _Atomic uintnat atomic_uintnat;
+typedef _Atomic intnat atomic_intnat;
+#endif
 
 #ifndef NULL
 #define NULL 0
@@ -619,7 +630,7 @@ CAMLextern int caml_read_directory(char_os * dirname,
 
 /* runtime message flags. Settable with v= in OCAMLRUNPARAM */
 
-extern _Atomic uintnat caml_verb_gc;
+extern atomic_uintnat caml_verb_gc;
 
 /* Bits which may be set in caml_verb_gc. The quotations are from the
  * OCaml manual. */

